snuba/clickhouse/formatter/query.py

```diff
@@ -1,174 +1,251 @@
+    composite.
 
     TODO: Remove this method entirely and move the sampling logic
     into a query processor.
     """
 
-    if settings.get_turbo() and not query.get_from_clause().sampling_rate:
-        query.set_from_clause(
-            replace(
-                query.get_from_clause(), sampling_rate=snuba_settings.TURBO_SAMPLE_RATE
+    if isinstance(query, Query):
+        if settings.get_turbo() and not query.get_from_clause().sampling_rate:
+            query.set_from_clause(
+                replace(
+                    query.get_from_clause(),
+                    sampling_rate=snuba_settings.TURBO_SAMPLE_RATE,
+                )
             )
-        )
-
-    return format_simple_query(query)
+        return FormattedQuery(_format_single_query_content(query))
+    else:
+        return FormattedQuery(_format_composite_query_content(query))
 
 
 def format_data_source(data_source: DataSource) -> FormattedNode:
     """
     Builds a FormattedNode out of any type of DataSource object. This
     is called when formatting the FROM clause of every nested query
     in a Composite query.
+
+    Remark: The ideal way to structure this would be a visitor just
+    like we did for the AST expression but this is not possible
+    without introducing either circular dependencies or dropping
+    type checking since the visitor and the classes to visit cannot
+    be in the same module as of now and in a visitor pattern they would
+    have to be mutually dependent.
     """
     if isinstance(data_source, Table):
         return format_table(data_source)
     elif isinstance(data_source, Query):
-        return format_simple_query(data_source)
+        return FormattedSubQuery(_format_single_query_content(data_source))
     elif isinstance(data_source, JoinClause):
         return data_source.accept(JoinFormatter())
     elif isinstance(data_source, CompositeQuery):
-        raise NotImplementedError("Composite queries not yet implemented")
+        return FormattedSubQuery(_format_composite_query_content(data_source))
     else:
         raise NotImplementedError(f"Unsupported data source {type(data_source)}")
 
 
 def format_table(data_source: Table) -> StringNode:
+    """
+    Formats a simple table data source.
+    """
+
     final = " FINAL" if data_source.final else ""
     sample = f" SAMPLE {data_source.sampling_rate}" if data_source.sampling_rate else ""
     return StringNode(f"{data_source.table_name}{final}{sample}")
 
 
-def format_simple_query(query: Query) -> FormattedQuery:
+def _format_single_query_content(query: Query) -> Sequence[FormattedNode]:
     """
-    Formats a Clickhouse Query from the AST representation into an
-    intermediate structure that can either be serialized into a string
-    (for clickhouse) or extracted as a sequence (for logging and tracing).
+    Produces the content for a simple physical query that references
+    a single table.
+    """
+    return _format_query_content(
+        [
```

Comment:
This structure is my best attempt to avoid duplication between simple and composite query. There is an annoying difference between the two (composite queries do noto have prewhere), thus we cannot just simply reuse the same method. This allows us to break it down step by step and reuse the steps.

Reply:
This seems like a lot of boiler plate just to handle that one case. If you had one format function that accepted both types, could you do:

```
prewhere = None
if not isinstance(query, CompositeQuery):
    prewhere = ...
```

That seems cleaner than having all of this listed out twice.

Reply:
You are right, your solution seems smarter. Will try.

Reply:
Removed some boilerplate.